File: apps/api/src/veupath_chatbot/domain/strategy/plan.py
"""Strategy plan data models.

These models represent the structured strategy plan that the LLM agent builds,
the user inspects/edits via an interactive UI, and the system executes.  They
are pure data models with no I/O — part of the domain layer.
"""

from datetime import UTC, datetime
from enum import StrEnum
from uuid import uuid4

from pydantic import Field

from veupath_chatbot.platform.pydantic_base import CamelModel
from veupath_chatbot.platform.types import JSONValue

# ── Enums ───────────────────────────────────────────────────────────


class PlanStatus(StrEnum):
    """Lifecycle status of a StrategyPlan."""

    DRAFT = "draft"
    PRESENTED = "presented"
    APPROVED = "approved"
    EXECUTING = "executing"
    COMPLETE = "complete"
    FAILED = "failed"


class StepStatus(StrEnum):
    """Readiness status of an individual PlannedStep."""

    READY = "ready"
    NEEDS_DISCOVERY = "needs_discovery"
    NEEDS_USER_INPUT = "needs_user_input"
    EXECUTING = "executing"
    COMPLETE = "complete"
    FAILED = "failed"


class StepType(StrEnum):
    """Structural type of a PlannedStep."""

    LEAF = "leaf"
    COMBINE = "combine"
    TRANSFORM = "transform"


class ParamStatus(StrEnum):
    """Resolution status of a PlannedParameter."""

    SET = "set"
    DEFAULT = "default"
    NEEDS_DISCOVERY = "needs_discovery"
    NEEDS_USER_INPUT = "needs_user_input"
    USER_SET = "user_set"


# ── Helpers ─────────────────────────────────────────────────────────


def _generate_plan_id() -> str:
    """Generate a plan ID: ``plan_<12 hex chars>``."""
    return f"plan_{uuid4().hex[:12]}"


def _utc_now() -> datetime:
    """Return the current UTC datetime."""
    return datetime.now(UTC)


# ── Data Models ─────────────────────────────────────────────────────


class PlannedParameter(CamelModel):
    """A parameter within a planned step, with its resolution status."""

    name: str
    display_name: str
    param_type: str
    value: JSONValue | None = None
    status: ParamStatus
    required: bool
    description: str | None = None
    constraints: dict[str, JSONValue] | None = None
    depends_on: list[str] = Field(default_factory=list)
    vocabulary_summary: str | None = None
    question: str | None = None
    options: list[str] | None = None
    rationale: str | None = None


class QuestionOption(CamelModel):
    """One option presented to the user for a UserQuestion."""

    label: str
    description: str = ""
    pros: list[str] = Field(default_factory=list)
    cons: list[str] = Field(default_factory=list)
    recommended: bool = False


class UserQuestion(CamelModel):
    """A question the system needs to ask the user to resolve ambiguity."""

    id: str
    question: str
    context: str = ""
    related_step: str | None = None
    related_param: str | None = None
    options: list[QuestionOption] | None = None
    answer: JSONValue | None = None


class PlannedStep(CamelModel):
    """A single step in a strategy plan."""

    id: str
    search_name: str
    display_name: str
    record_type: str = "transcript"
    rationale: str = ""
    step_type: StepType
    status: StepStatus
    parameters: dict[str, PlannedParameter] = Field(default_factory=dict)
    operator: str | None = None
    expected_count: int | None = None
    actual_count: int | None = None
    wdk_step_id: int | None = None
    graph_step_id: str | None = None


class PlannedConnection(CamelModel):
    """A directed edge between two planned steps."""

    from_step: str
    to_step: str
    input_type: str = "primary"
    operator: str | None = None


class StrategyPlan(CamelModel):
    """The full strategy plan: steps, connections, questions, metadata."""

    id: str = Field(default_factory=_generate_plan_id)
    title: str
    description: str
    rationale: str
    status: PlanStatus = PlanStatus.DRAFT
    version: int = 1
    steps: list[PlannedStep]
    connections: list[PlannedConnection]
    questions: list[UserQuestion] = Field(default_factory=list)
    uncertainties: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=_utc_now)
    updated_at: datetime = Field(default_factory=_utc_now)

    def steps_in_dependency_order(self) -> list[PlannedStep]:
        """Return steps topologically sorted so dependencies come first.

        Leaf steps (no incoming connections) come first, then combine/transform
        steps whose inputs are already resolved.  Raises ``ValueError`` if
        the connection graph contains a cycle.
        """
        step_by_id = {s.id: s for s in self.steps}

        # Build adjacency: to_step depends on from_step
        dependents: dict[str, set[str]] = {s.id: set() for s in self.steps}
        for conn in self.connections:
            if conn.to_step in dependents and conn.from_step in step_by_id:
                dependents[conn.to_step].add(conn.from_step)

        # Kahn's algorithm
        ordered: list[PlannedStep] = []
        ready = [sid for sid, deps in dependents.items() if not deps]

        while ready:
            sid = ready.pop(0)
            ordered.append(step_by_id[sid])
            for other, deps in dependents.items():
                if sid in deps:
                    deps.discard(sid)
                    if not deps:
                        ready.append(other)

        if len(ordered) != len(self.steps):
            msg = "Cycle detected in plan connections"
            raise ValueError(msg)

        return ordered

File: apps/web/src/lib/types/plan.ts
import { z } from "zod";

export type PlanStatus = "draft" | "presented" | "approved" | "executing" | "complete" | "failed";
export type StepStatus = "ready" | "needs_discovery" | "needs_user_input" | "executing" | "complete" | "failed";
export type StepType = "leaf" | "combine" | "transform";
export type ParamStatus = "set" | "default" | "needs_discovery" | "needs_user_input" | "user_set";

// ── Zod schemas (SSOT — interfaces are inferred from these) ────────

export const PlannedParameterSchema = z.object({
  name: z.string(),
  displayName: z.string(),
  paramType: z.string(),
  value: z.unknown().nullable(),
  status: z.enum(["set", "default", "needs_discovery", "needs_user_input", "user_set"]),
  required: z.boolean(),
  description: z.string().nullable(),
  constraints: z.record(z.string(), z.unknown()).nullable(),
  dependsOn: z.array(z.string()).default([]),
  vocabularySummary: z.string().nullable().default(null),
  question: z.string().nullable().default(null),
  options: z.array(z.string()).nullable().default(null),
  rationale: z.string().nullable().default(null),
});

export const QuestionOptionSchema = z.object({
  label: z.string(),
  description: z.string().default(""),
  pros: z.array(z.string()).default([]),
  cons: z.array(z.string()).default([]),
  recommended: z.boolean().default(false),
});

export const UserQuestionSchema = z.object({
  id: z.string(),
  question: z.string(),
  context: z.string().default(""),
  relatedStep: z.string().nullable().default(null),
  relatedParam: z.string().nullable().default(null),
  options: z.array(QuestionOptionSchema).nullable().default(null),
  answer: z.unknown().nullable().default(null),
});

export const PlannedStepSchema = z.object({
  id: z.string(),
  searchName: z.string(),
  displayName: z.string(),
  recordType: z.string().default("transcript"),
  rationale: z.string().default(""),
  stepType: z.enum(["leaf", "combine", "transform"]),
  status: z.enum(["ready", "needs_discovery", "needs_user_input", "executing", "complete", "failed"]),
  parameters: z.record(z.string(), PlannedParameterSchema).default({}),
  operator: z.string().nullable().default(null),
  expectedCount: z.number().nullable().default(null),
  actualCount: z.number().nullable().default(null),
  wdkStepId: z.number().nullable().default(null),
  graphStepId: z.string().nullable().default(null),
});

export const PlannedConnectionSchema = z.object({
  fromStep: z.string(),
  toStep: z.string(),
  inputType: z.string().default("primary"),
  operator: z.string().nullable().default(null),
});

export const InteractivePlanSchema = z.object({
  id: z.string(),
  title: z.string(),
  description: z.string(),
  rationale: z.string(),
  status: z.enum(["draft", "presented", "approved", "executing", "complete", "failed"]).default("draft"),
  version: z.number().default(1),
  steps: z.array(PlannedStepSchema).default([]),
  connections: z.array(PlannedConnectionSchema).default([]),
  questions: z.array(UserQuestionSchema).default([]),
  uncertainties: z.array(z.string()).default([]),
  createdAt: z.string().default(""),
  updatedAt: z.string().default(""),
});

// ── Inferred types (no manual interface duplication) ────────────────

export type PlannedParameter = z.infer<typeof PlannedParameterSchema>;
export type QuestionOption = z.infer<typeof QuestionOptionSchema>;
export type UserQuestion = z.infer<typeof UserQuestionSchema>;
export type PlannedStep = z.infer<typeof PlannedStepSchema>;
export type PlannedConnection = z.infer<typeof PlannedConnectionSchema>;
export type InteractivePlan = z.infer<typeof InteractivePlanSchema>;

export interface PlanParameterEdit {
  stepId: string;
  paramName: string;
  newValue: unknown;
}

export interface PlanQuestionAnswer {
  questionId: string;
  answer: unknown;
}

export interface PlanActionRequest {
  action: "approve" | "reject" | "suggest_changes";
  planId: string;
  traceId?: string | null;
  messageGroupId?: string | null;
  source?: string | null;
  paramEdits?: PlanParameterEdit[];
  answers?: PlanQuestionAnswer[];
}

export interface PlanInteractionMetadata {
  type: "plan_interaction";
  planId: string;
  action: "approve" | "update_parameter" | "answer_question" | "reject" | "suggest_changes";
  data: {
    stepId?: string;
    paramName?: string;
    newValue?: unknown;
    questionId?: string;
    answer?: unknown;
    suggestion?: string;
    paramEdits?: Array<{ stepId: string; paramName: string; newValue: unknown }>;
    answers?: Array<{ questionId: string; answer: unknown }>;
  };
}
